File: sorting/quicksort.java
public class quicksort {

    // Method to partition the array and return the index of the pivot
    public static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];  // Choose the last element as pivot
        int i = (low - 1);  // Index of the smaller element
        
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        
        // Swap the pivot element to its correct position
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;  // Return the pivot index
    }

    // The main Quick Sort function
    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            // Partition the array and get the pivot index
            int pi = partition(arr, low, high);

            // Recursively sort elements before and after partition
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    public static void main(String[] args) {
        int[] arr = {9, 7, 5, 11, 12, 2, 14, 3, 10, 6};  // Sample array

        System.out.println("Original array:");
        printArray(arr);  // Print original array

        quickSort(arr, 0, arr.length - 1);  // Sort the array using Quick Sort

        System.out.println("Sorted array:");
        printArray(arr);  // Print sorted array
    }

    // Helper method to print an array
    public static void printArray(int[] arr) {
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();
    }
}

File: sorting/quicksort.py
# quicksort is a recursive algorithm - follows divide and conquer strategy by selecting pivot
# TC - best and average O(n logn)
# worst - O(n^2)
# as quicksort is recursive it maintains a recursion stack which requires additional memory
# space complecit - worst: O(n), avg: O(logn)

# The iterative approach is more efficient than recursive.

# Recursion in Quick Sort is generally efficient for most cases, but it's important to understand 
# its limitations (like the worst case). If your concern is worst-case performance, 
# you might want to consider using Merge Sort or hybrid sorting algorithms like Timsort. 

def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    else:
        pivot = arr[-1]
        smaller, equal, larger = [], [], []

        for num in arr:
            if num < pivot:
                smaller.append(num)
            elif num == pivot:
                equal.append(num)
            else:
                larger.append(num)

        return quick_sort(smaller) + equal + quick_sort(larger)


if __name__ == "__main__":
    arr = [9,7,5,11,12,2,14,3,10,6]
    print("Original array : ", arr)
    sorted_arr = quick_sort(arr)
    print("Sorted array : ", sorted_arr)

